src/filter.rs

```diff
@@ -0,0 +1,322 @@
+//! Interfaces for adding custom filter to libgit2
+
+use libc::{c_int, c_char, c_void, size_t};
+use std::{mem, slice, ptr};
+use {raw, call, util, Error, panic};
+use std::ffi::CString;
+use std::path::Path;
+use raw::{git_filter, git_writestream};
+
+#[repr(C)]
+struct Filter {
+    filter: git_filter,
+    attrs: Option<CString>,
+    instance: Box<StreamFilter>,
+}
+
+#[repr(C)]
+struct Stream {
+    parent: git_writestream,
+    handler: Box<WriteStream>,
+}
+
+/// Wrapper of the raw git_writestream object, representing the input
+/// of next filter of the chain
+pub struct RawWriteStream {
+    stream: *mut git_writestream,
+}
+
+/// A trait representing operations to handle filter data stream.
+///
+/// These callbacks are typically called by libgit2.
+pub trait WriteStream : Send + 'static {
+    /// New stream data, Implementations should process the buffer data
+    /// and then write the output to the next stream if needed.
+    fn write(&mut self, buf: &[u8]) -> Result<(), Error>;
+
+    /// Close stream. Implementations should do finalizing work and then
+    /// close the next stream.
+    fn close(&mut self) -> Result<(), Error>;
+}
+
+/// Information to construct a new stream filter.
+pub struct FilterSource {
+    raw: *const raw::git_filter_source
```

Comment:
I don't think that this owns the pointer so this'll want a `<'a>` lifetime parameter on it somewhere I believe.